driver-core/src/main/com/mongodb/connection/InternalStreamConnection.java

```diff
@@ -181,90 +189,230 @@ public boolean isClosed() {
+            close();
+            readingPhase.getAndSet(localLatch).countDown();
+        } finally {
+            readerLock.unlock();
+        }
+
+        while (true) {
+            if (isClosed()) {
+                if (exceptionThatPrecededStreamClosing != null) {
+                    throw exceptionThatPrecededStreamClosing;
                 } else {
-                    throw MongoException.fromThrowable(t);
+                    throw new MongoSocketClosedException("Socket has been closed", getServerAddress());
                 }
             }
-            return response.getResult();
+            ResponseBuffers myResponse = messages.remove(responseTo);
+            if (myResponse != null) {
+                connectionListener.messageReceived(new ConnectionMessageReceivedEvent(getId(),
+                                                                                      myResponse.getReplyHeader().getResponseTo(),
+                                                                                      myResponse.getReplyHeader().getMessageLength()));
+                return myResponse;
+            }
+
+            try {
+                localLatch.await();
+            } catch (InterruptedException e) {
+                throw new MongoInterruptedException("Interrupted while reading from stream", e);
+            }
+
+            localLatch = readingPhase.get();
         }
     }
 
     @Override
     public void sendMessageAsync(final List<ByteBuf> byteBuffers, final int lastRequestId, final SingleResultCallback<Void> callback) {
-        notNull("open", stream, callback);
+        notNull("stream is open", stream, callback);
+
+        if (isClosed()) {
+            callback.onResult(null, new MongoSocketClosedException("Can not read from a closed socket", getServerAddress()));
+        }
+
         if (LOGGER.isTraceEnabled()) {
-            LOGGER.trace(format("Queuing send message: %s. ([%s] %s)", lastRequestId, getId(), serverId));
+            LOGGER.trace(format("Queuing send message: %s. ([%s])", lastRequestId, getId()));
+        }
+
+        SendMessageRequest sendMessageRequest = new SendMessageRequest(byteBuffers, lastRequestId, errorHandlingCallback(callback, LOGGER));
+
+        boolean mustWrite = false;
+        writerLock.lock();
+        try {
+            if (isWriting) {
+                writeQueue.add(sendMessageRequest);
+            } else {
+                isWriting = true;
+                mustWrite = true;
+            }
+
+        } finally {
+            writerLock.unlock();
+        }
+
+        if (mustWrite) {
+            writeAsync(sendMessageRequest);
         }
-        writeQueue.add(new SendMessageAsync(byteBuffers, lastRequestId, errorHandlingCallback(callback, LOGGER)));
-        processPendingWrites();
+    }
+
+    private void writeAsync(final SendMessageRequest request) {
+        stream.writeAsync(request.getByteBuffers(), new AsyncCompletionHandler<Void>() {
```

Comment:
An isClosed() check is problematic because there is no synchronization across readers and writers on the isClosed flag, so there will always be a race when checking it.

However, I think this will work out ok.  the close() method does not null out the stream member variable, so worst case thread 2 calls stream.writeAsync and the callback gets an exception from the stream saying that it's been closed.